[RISCV][TII] Add and use new hook to optimize/canonicalize instructions after MachineCopyPropagation

asb opened the pull request:
PR #136875 was posted as a draft PR that handled a subset of these cases, using the CompressPat mechanism. The consensus from that discussion (and a conclusion I agree with) is that it would be beneficial doing this optimisation earlier on, and in a way that isn't limited just to cases that can be handled by instruction compression.

The most common source for instructions that can be optimized/canonicalized in this way is through tail duplication in MachineBlockPlacement followed by machine copy propagation. For RISC-V, choosing a more canonical instruction allows it to be compressed when it couldn't be before. There is the potential that it would make other MI-level optimisations easier.

This modifies ~910 instructions across an llvm-test-suite build including SPEC2017, targeting rva22u64. Looking at the diff, it seems there's room for eliminating instructions or further propagating after this.

Coverage of instructions is based on observations from a script written to find redundant or improperly canonicalized instructions (though I aim to support all instructions in a 'group' at once, e.g. MUL* even if I only saw some variants of MUL in practice).

---
The obvious thing to bikeshed is the name of the hook. I do worry based on the name, it might be assumed to be a more generic form of other hooks like `optimizeCompareInstr` or `optimizeSelect`. Alternate names welcome. `optimizeMutatedInstruction` might capture the idea that this is intended to be run after you mutate the operands of an instruction in order to optimize/canonicalize it to a "better" one if possible.

This ended up covering rather a lot more instructions than I originally thought, so I'd appreciate extra eyes checking there are no silly mistakes in the matched patterns.

asb committed:
[RISCV][TII] Add and use new hook fo optmize/canonicalize instructions after MachineCopyPropagation

PR #136875 was posted as a draft PR that handled a subset of these
cases, using the CompressPat mechanism. The consensus from that
discussion (and a conclusion I agree with) is that it would be
beneficial doing this optimisation earlier on, and in a way that isn't
limited just to cases that can be handled by instruction compression.

The most common source for instructions that can be
optimized/canonicalized in this way is through tail duplication followed
by machine copy propagation. For RISC-V, choosing a more canonical
instruction allows it to be compressed when it couldn't be before. There
is the potential that it would make other MI-level optimisations easier.

This modifies ~910 instructions across an llvm-test-suite build
including SPEC2017, targeting rva22u64.

Coverage of instructions is based on observations from a script written
to find redundant or improperly canonicalized instructions (though I
aim to support all instructions in a 'group' at once, e.g. MUL* even if
I only saw some variants of MUL in practice).

asb committed:
Add changes missed in first push

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,209 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+      MI.setDesc(get(RISCV::ADDI));
+      return true;
+    }
+    // add.uw rd, zero, rs => add.uw rd, rs, zero (canonical zext.w)
```
This doesn't look right. Only the first operand of add.uw is zero extended. `add.uw rd, zero, rs` would be equivalent to `mv rd, rs`.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,209 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+        MI.setDesc(get(RISCV::ADDI));
+        return true;
+      }
+      // xor rd, rs, rs => li rd, rs, 0
```
`li rd, rs, 0` isn't a valid instruction. The code does `addi rd, rs, 0` which is `mv rd, rs`, but `xor rd, rs, rs1` is equivalent to `li rd, 0`

asb replied:
Thanks. Now fixed (I was also missing a test here).

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,209 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+      // Normalize:
+      // [x]or rd, zero, rs => [x]or rd, rs, zero
+      if (MI.getOperand(1).getReg() == RISCV::X0) {
+        MachineOperand MO1 = MI.getOperand(1);
```
I wonder if using `commuteInstruction` would be more obvious what the code is doing?

asb replied:
Good idea, I've used commuteInstruction here and for addw

arsenm commented on llvm/include/llvm/CodeGen/TargetInstrInfo.h:
```diff
@@ -510,6 +510,16 @@ class TargetInstrInfo : public MCInstrInfo {
+  /// MachineCopyPropagation, where their mutation of the MI operands may
+  /// expose opportunities to convert the instruction to a simpler form (e.g.
+  /// a load of 0).
+  virtual bool optimizeInstruction(MachineInstr &MI) const { return false; }
```
Name is way too general for a very narrowly applied optimization. Can't this just go in a separate post-RA pass

preames replied:
@asb Can you say a bit about why this needs to be in MCP?  As opposed to just after MCP?  Does doing this in the process of copy propagation expose additional copies?  (Not implausible, but do you actually see this?)

topperc replied:
The problem this patch is solving is probably not unique to RISC-V. Having the ability for other targets to do this type of canonicalization seems like a good idea.

Are we suggesting a separate target independent pass using this hook? Or a target specific pass?

asb replied:
Craig's comment captures well the reason why I looked to do this in MCP. it seems like a transformation that's generally useful to other targets, and by adding and using a TII hook it's easy to reuse if we find any other target-specific or generic passes that might benefit (though as I say in the summary, all of the instances of these instructions I can find come from MCP). It could be done in another pass, but it felt like a simple change to have MCP "check its own work" might be cleaner than having yet another pass.

This specific change doesn't induce additional copy propagation, but looking at the generated diffs there are cases where you would expect copy propagation should be able to run again. I nod to this in the description above, but as this change alone is an improvement on before I leave that to a followup.

e.g. this snippet from parest:
```diff
@@ -2753,11 +2753,11 @@
        bnez    s11, .LBB0_353
 .LBB0_357:                              #   in Loop: Header=BB0_310 Depth=1
        li      a2, 0
-       srliw   a5, zero, 31
-       slli    a3, zero, 33
+       li      a5, 0
+       li      a3, 0
        slli    a5, a5, 15
        srli    a3, a3, 56
-       and     a4, zero, a1
+       li      a4, 0
        bgeu    s8, a3, .LBB0_354
```

The slli/srli should be removable (canonicalised to loadimm of 0, and then redundant with the previous `li`). I haven't yet stepped through to see if this is a matter of doing another iteration in MCP or if there's some other barrier that prevents current MCP from handling the case.

arsenm replied:
This is something that can constant fold, but didn't earlier? At this point I would hope that would have been cleaned long before this

asb replied:
Oh I see the cause of confusion - this is indeed cleaning up after tail duplication - the use of the TailDuplicator utility class (llvm/lib/CodeGen/TailDuplicator.cpp) in MachineBlockPlacement.

Here is a roughly reduced example representing the above snippet:
```
; ModuleID = '<stdin>'
source_filename = "<stdin>"
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
target triple = "riscv64-unknown-linux-gnu"

define i64 @ham(i1 %arg, i32 %arg1) {
bb:
  br label %bb2

bb2:                                              ; preds = %bb15, %bb
  %and = and i32 0, 8388607
  br i1 %arg, label %bb4, label %bb3

bb3:                                              ; preds = %bb2
  br label %bb4

bb4:                                              ; preds = %bb3, %bb2
  %phi = phi i32 [ %arg1, %bb3 ], [ 0, %bb2 ]
  %lshr = lshr i32 %phi, 16
  %and5 = and i32 %lshr, 32768
  %lshr6 = lshr i32 %phi, 23
  %and7 = and i32 %lshr6, 255
  %and8 = and i32 %phi, 8388607
  %icmp = icmp ult i32 %and7, 113
  br i1 %icmp, label %bb9, label %bb10

bb9:                                              ; preds = %bb4
  %or = or i32 %and8, %and5
  %trunc = trunc i32 %or to i16
  br label %bb15

bb10:                                             ; preds = %bb4
  br i1 %arg, label %bb11, label %bb13

bb11:                                             ; preds = %bb10
  %trunc12 = trunc i32 %and5 to i16
  br label %bb15

bb13:                                             ; preds = %bb10
  %trunc14 = trunc i32 %and8 to i16
  br label %bb15

bb15:                                             ; preds = %bb13, %bb11, %bb9
  %phi16 = phi i16 [ %trunc, %bb9 ], [ %trunc12, %bb11 ], [ %trunc14, %bb13 ]
  %trunc17 = trunc i16 %phi16 to i8
  store i8 %trunc17, ptr null, align 1
  br label %bb2
}
```
If you run that through `llc -O3` you'll see the tail duplication happening as part of MachineBlockPlacement and then MCP runs.

So you have a block:
```
bb.4.bb4:
; predecessors: %bb.3, %bb.2
  successors: %bb.5(0x40000000), %bb.6(0x40000000); %bb.5(50.00%), %bb.6(50.00%)
  liveins: $x10, $x11, $x12, $x13, $x15
  renamable $x14 = SRLIW renamable $x15, 31
  renamable $x16 = SLLI renamable $x15, 33
  renamable $x14 = SLLI killed renamable $x14, 15
  renamable $x16 = SRLI killed renamable $x16, 56
  renamable $x15 = AND killed renamable $x15, renamable $x12
  BLTU renamable $x13, killed renamable $x16, %bb.6
```

After tail duplication is applied in MBP you get:
```
bb.2:
; predecessors: %bb.1
  successors: %bb.5(0x40000000), %bb.6(0x40000000); %bb.5(50.00%), %bb.6(50.00%)
  liveins: $x10, $x11, $x12, $x13
  $x15 = ADDI $x0, 0
  renamable $x14 = SRLIW renamable $x15, 31
  renamable $x16 = SLLI renamable $x15, 33
  renamable $x14 = SLLI killed renamable $x14, 15
  renamable $x16 = SRLI killed renamable $x16, 56
  renamable $x15 = AND killed renamable $x15, renamable $x12
  BGEU renamable $x13, killed renamable $x16, %bb.5
  PseudoBR %bb.6

bb.3.bb3:
; predecessors: %bb.1
  successors: %bb.5(0x40000000), %bb.6(0x40000000); %bb.5(50.00%), %bb.6(50.00%)
  liveins: $x10, $x11, $x12, $x13
  $x15 = ADDI renamable $x11, 0
  renamable $x14 = SRLIW renamable $x15, 31
  renamable $x16 = SLLI renamable $x15, 33
  renamable $x14 = SLLI killed renamable $x14, 15
  renamable $x16 = SRLI killed renamable $x16, 56
  renamable $x15 = AND killed renamable $x15, renamable $x12
  BGEU renamable $x13, killed renamable $x16, %bb.5
```
Where obviously bb.2 can be cleaned up which MCP does to a certain extent.

asb committed:
Fix errors in some patterns and remove PACK/PACKW

asb commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,209 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+  case RISCV::PACK:
+  case RISCV::PACKW:
+      // sub rd, rs, zero => addi rd, rs, 0
+      // pack[w] rd, rs, zero => addi rd, rs, zero
```
As always, thank you for your close attention to detail and sorry that I've had to rely on it! I'm not quite sure how I ended up writing these pack patterns - I've removed for now.

preames commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,207 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+  case RISCV::SH3ADD_UW:
+      // shNadd[.uw] rd, zero, rs => addi rd, rs, 0
+      if (MI.getOperand(1).getReg() == RISCV::X0) {
+        MI.removeOperand(1);
```
Remove/Add is an odd idiom here.  Maybe ChangeToImm instead?

topperc replied:
It's removing X0 which shifts over rs, and then adding 0 at the end. So ChangeToImm won't work.

preames commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,207 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+      if (MI.getOperand(2).getReg() == RISCV::X0) {
+        MI.removeOperand(2);
+        unsigned Opc = MI.getOpcode();
+        if (Opc == RISCV::SH1ADD_UW || Opc == RISCV::SH2ADD_UW || Opc == RISCV::SH3ADD_UW) {
```
Maybe just a NewOpc variable with shared mutation?

preames commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,207 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+      // and rd, zero, rs => li rd, 0
+      // mul* rd, rs, zero => li rd, 0
+      // mul* rd, zero, rs => li rd, 0
+      if (MI.getOperand(1).getReg() == RISCV::X0) {
```
Why two ifs instead of a compound comparison?

asb replied:
I've restructured this. Thanks.

preames commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,207 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+        return true;
+      }
+      break;
+  case RISCV::ORI:
```
For code organization, maybe keep the OR and ORI closer together?  (Used as an illustration for all the pairs, not just these two specifically.)

asb replied:
I've reordered to keep non-imm and imm forms together, and more generally have the mv-like forms higher up the list than the loadimm forms.

preames commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,207 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+      }
+      break;
+  case RISCV::SLTIU:
+      // seqz rd, zero => li rd, 1
```
Comment and code out of sync?

topperc replied:
The comment is correct if you take into account that seqz and li are both aliases.

`seqz rd, zero` is `sltiu rd, zero, 1`
`li rd, 1` is `addi rd, zero, 1`

asb replied:
As Craig says, the comment does match what is being done, but looking through I've been inconsistent about using pseudoinstructions or not in the comments. Given the precise form of the instruction is the important thing rather than its semantic effect, I've changed to always just use the raw instruction form. Hopefully it's a lot easier to follow now.

preames replied:
Yep, thanks, much easier to read now.

preames commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,207 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+      break;
+  case RISCV::SLTU:
+  case RISCV::ADD_UW:
+    // snez rd, zero => li rd, 0
```
Again - comment != code.

topperc replied:
`snez rd, rs` is an alias for `sltu rd, zero, rs` so `snez rd, zero, zero` is `sltu rd zero, zero`.

asb replied:
As above, I've changed all these comments to avoid use of pseudoinstructions. Marking this thread as resolved as if there's further thoughts on that change we can discuss in the sibling thread.

preames commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -2344,6 +2344,21 @@ static unsigned getSHXADDShiftAmount(unsigned Opc) {
 
+// Returns the shift amount from a SHXADD.UW instruction. Returns 0 if the
+// instruction is not a SHXADD.UW.
+static unsigned getSHXADDUWShiftAmount(unsigned Opc) {
```
I'd probably just merge this with the previous routine.  Reading the using code, I didn't even spot the difference in the name because I'd assumed it to be one routine.

topperc replied:
I almost made the same comment, but the original caller of the other function uses the return value to know if the instruction is a SHXADD. That caller would be incorrect for SHXADD_UW.

preames commented on llvm/lib/CodeGen/MachineCopyPropagation.cpp:
```diff
@@ -867,6 +867,12 @@ void MachineCopyPropagation::forwardUses(MachineInstr &MI) {
          make_range(Copy->getIterator(), std::next(MI.getIterator())))
       KMI.clearRegisterKills(CopySrcReg, TRI);
 
+    // Attempt to canonicalize/optimize the instruction now its arguments have
```
The placement here is unsound.  You're in a loop which is operating over the instructions operands!

asb replied:
Ouch yes, I'd meant to place it outside the loop. Thanks for catching - now fixed.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,207 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+  case RISCV::ADDW:
+      // Normalize:
+      // addw rd, zero, rs => addw rd, rs, zero
+      if (MI.getOperand(1).getReg() == RISCV::X0) {
```
ADDW does have the commutable flag. It's used by MachineCSE.

asb replied:
I've now switched to using commuteInstruction here.

asb committed:
Use commuteInstruction helper

asb committed:
(cherry-pick) [RISCV][NFC] Add missed // clang-format on

clang-format was turned off for the defines, but there was no matching
`// clang-format on` comment at the end. Ran into this in #137973

https://github.com/llvm/llvm-project/commit/3376071a24b772446d2f2241e4370b6d0cba6947

asb committed:
Move optimizeInstructoin to after the preprocessor undefs

asb committed:
Reformat now issue with clang-format being disabled was fixed

asb commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,207 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+        MachineOperand MO1 = MI.getOperand(1);
+        MI.removeOperand(1);
+        MI.addOperand(MO1);
+      }
```
We do need to be careful of that, but note that the precondition for commuting the operands means that the following conditional is always true. i.e. it should not be possible to commute the instruction and return false.

I thought about a comment along these lines, but failed to come up with anything that didn't make it seem more mysterious or complicated than the reality. Do you think this needs explanation?

preames replied:
You have "Normalize" as a comment, maybe "Normalize (so we hit the next if-clause)".  Or something along those lines.  I didn't spot the fallthrough on first glance, though I agree it's easier to see after the use of commuteInstruction.

asb commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3850,6 +3865,207 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+  switch (MI.getOpcode()) {
+  default:
+    break;
+  case RISCV::OR:
```
Just to make sure I follow, are you suggesting changing from what I have now:
```cpp
  case RISCV::OR:
  case RISCV::XOR:
    // Normalize:
    // [x]or rd, zero, rs => [x]or rd, rs, zero
    if (MI.getOperand(1).getReg() == RISCV::X0)
      commuteInstruction(MI);
    // [x]or rd, rs, zero => addi rd, rs, 0
    if (MI.getOperand(2).getReg() == RISCV::X0) {
      MI.getOperand(2).ChangeToImmediate(0);
      MI.setDesc(get(RISCV::ADDI));
      return true;
    }
    // xor rd, rs, rs => li rd, 0
    if (MI.getOpcode() == RISCV::XOR &&
        MI.getOperand(1).getReg() == MI.getOperand(2).getReg()) {
      MI.getOperand(1).setReg(RISCV::X0);
      MI.getOperand(2).ChangeToImmediate(0);
      MI.setDesc(get(RISCV::ADDI));
      return true;
    }
    break;
```

To:
```cpp
  case RISCV::OR:
    // Normalize:
    // or rd, zero, rs => or rd, rs, zero
    if (MI.getOperand(1).getReg() == RISCV::X0)
      commuteInstruction(MI);
    // or rd, rs, zero => addi rd, rs, 0
    if (MI.getOperand(2).getReg() == RISCV::X0) {
      MI.getOperand(2).ChangeToImmediate(0);
      MI.setDesc(get(RISCV::ADDI));
      return true;
    }
    break;
  case RISCV::XOR:
    // Normalize:
    // xor rd, zero, rs => xor rd, rs, zero
    if (MI.getOperand(1).getReg() == RISCV::X0)
      commuteInstruction(MI);
    // xor rd, rs, zero => addi rd, rs, 0
    if (MI.getOperand(2).getReg() == RISCV::X0) {
      MI.getOperand(2).ChangeToImmediate(0);
      MI.setDesc(get(RISCV::ADDI));
      return true;
    }
    // xor rd, rs, rs => li rd, 0
    if (MI.getOpcode() == RISCV::XOR &&
        MI.getOperand(1).getReg() == MI.getOperand(2).getReg()) {
      MI.getOperand(1).setReg(RISCV::X0);
      MI.getOperand(2).ChangeToImmediate(0);
      MI.setDesc(get(RISCV::ADDI));
      return true;
    }
    break;
```

Happy to switch if you prefer, but at least to my eye it doesn't seem clear cut that this is better than before.

asb committed:
Consistently avoid use of pseudoinstructions in comments for transforms

asb committed:
Pull optimizeInstruction call outside of loop

asb committed:
Simplify and/mul* case as suggested in review

asb committed:
Re-order instructions in switch

reg-reg and reg-imm forms are now close to each other. In general we
handle the mv-like transforms first, then the loadimm ones.

preames commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3872,6 +3888,206 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+  case RISCV::SLTIU:
+    // sltiu rd, zero, 1 => addi rd, zero, 1
+    if (MI.getOperand(1).getReg() == RISCV::X0 &&
+        MI.getOperand(2).getImm() == 1) {
```
This should work for any non-zero immediate shouldn't it?

asb replied:
Thanks, I'd been checking for snez in my script and just transcribed that into a pattern. But it makes more sense to be more general. I now handle `sltiu rd, zero, NZC` and `sltiu rd, zero, 0` for completeness. Though there are zero instances in current codegen.

preames commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3872,6 +3888,206 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+    break;
+  case RISCV::SLL:
+  case RISCV::SRL:
+  case RISCV::SRA:
```
Possibly missing case: shift rd, rs, zero => addi rd, rs, 0  (for the non UW shifts only)

asb replied:
I've added that case. Scanning for this, there is only a single instance in the llvm-test-suite output (an `sll a0, a0, zero` in 7zip-benchmark). But I think it makes sense to be exhaustive for the opcodes that are matched (i.e. if we match an opcode it makes sense to implement all the obvious patterns, even if we skip some opcodes altogether).

preames reviewed (commented):
Another round of minor comments.  I generally think this is a good idea, subject to getting the interface clarified.

asb committed:
clang-format

arsenm commented on llvm/lib/CodeGen/MachineCopyPropagation.cpp:
```diff
@@ -870,6 +870,11 @@ void MachineCopyPropagation::forwardUses(MachineInstr &MI) {
+  // Attempt to canonicalize/optimize the instruction now its arguments have
+  // been mutated.
+  if (TII->optimizeInstruction(MI)) {
+    LLVM_DEBUG(dbgs() << "MCP: After optimizeInstruction: " << MI << "\n");
```
```suggestion
    LLVM_DEBUG(dbgs() << "MCP: After optimizeInstruction: " << MI);
```

Redundant newline

asb committed:
reorder comment to match code logic

asb committed:
generalise sltiu check to any non-zero immediate

asb committed:
Add sll/srl/sra mv-like case

asb committed:
handle sltiu rd, zero, 0

asb committed:
Add suggested comments for 'normalize' pre-transforms

asb committed:
Get rid of redundant newline

asb commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3872,6 +3888,206 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+  case RISCV::MULHSU:
+  case RISCV::MULHU:
+  case RISCV::MULW:
+    // and rd, rs, zero => addi rd, zero, 0
```
Done

asb commented:
Thank you everyone. I believe all of the comments on the details of the implementation have been addressed an are reflected in the pushed version of the code. I have tweaked the commit message to clarify that the tail duplication that produces the opportunity for these transformations is from tail duplication performed in MachineBlockPlacement.

I believe the open issues are:
* Hook naming. Some ideas
  * simplifyInstruction (on the basis that it applies basic simplifications - e.g. to li or mv)
  * optimizeMutatedInstruction (on the basis that it is an optimisation intended to be run after an instruction's operands have been mutated an instruction, which may present new opportunities for simplifying)
  * yourIdeaHere - honestly, I don't have strong views!
* Whether this makes sense as a tweak to MCP as proposed, or a separate pass
  * It would be possible to add a target specific pass or add to a new one, but it does seem like this is relevant to targets beyond just RISC-V and requires only a minimal tweak to MCP, with an interface that could be used from other target-specific passes that might need it. So if the preference is to _not_ go this route, I'd like to better understand the argument against.

preames reviewed (approved):
LGTM - I'm not really a fan of the generic name here, but don't have a better suggestion, and don't want this blocked indefinitely.

Once this lands, I will be posting a follow up patch which reorganizes MCP such that after this simplification, we can further propagate the copy propagation.  I feel the follow on with respect to that is key to motivating this direction.

topperc commented on llvm/lib/CodeGen/MachineCopyPropagation.cpp:
```diff
@@ -870,6 +870,11 @@ void MachineCopyPropagation::forwardUses(MachineInstr &MI) {
+  // Attempt to canonicalize/optimize the instruction now its arguments have
+  // been mutated.
+  if (TII->optimizeInstruction(MI)) {
+    LLVM_DEBUG(dbgs() << "MCP: After optimizeInstruction: " << MI);
```
Do we need to set `Changed = true` here or is guaranteed it was already set earlier?

asb replied:
Good question. It's not clear to me that Changed = true will _always_ have been set, so I've added an explicit assignment so it's obvious.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -3872,6 +3888,215 @@ MachineInstr *RISCVInstrInfo::commuteInstructionImpl(MachineInstr &MI,
+    }
+    break;
+  case RISCV::SUB:
+    // sub rd, rs, zero => addi rd, rs, 0
```
Does `add rd, rs, zero` never show up or we just aren't converting it to ADDI?

asb replied:
It does show up, but we do have a a CompressPat that produces c.mv for either case. So there's no real difference if C is enabled, and I've been running my script on build directories that have C enabled.

But there's no reason not to handle ADD here, and in the case of a target without the C extension it at least means you'll get a canonical `mv` in such cases which is nicer to read in disassembly if nothing else.

asb committed:
Set Changed=true if optimizeInstruction returned true in call from MCP

asb committed:
Cover ADD as well

asb committed:
Rename hook to simplifyInstruction

asb commented:
To my mind simplifyInstruction is at least slightly better, so I've changed to that in the absence of strong views on alternate names.

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q2-riscv-optimize-instruction-hook

asb committed:
Add comment to .mir file

asb committed:
Rename test

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q2-riscv-optimize-instruction-hook
